src/backend/distributed/commands/multi_copy.c

```diff
@@ -175,6 +179,227 @@ static void CopyAttributeOutText(CopyOutState outputState, char *string);
+
+		if (strncmp(defel->defname, "master_host", NAMEDATALEN) == 0)
+		{
+			masterNodeAddress->nodeName = defGetString(defel);
+		}
+		else if (strncmp(defel->defname, "master_port", NAMEDATALEN) == 0)
+		{
+			masterNodeAddress->nodePort = defGetInt32(defel);
+		}
+	}
+
+	return masterNodeAddress;
+}
+
+
+/*
+ * RemoveMasterOptions removes master node related copy options from the option
+ * list of the copy statement.
+ */
+static void
+RemoveMasterOptions(CopyStmt *copyStatement)
+{
+	List *newOptionList = NIL;
+	ListCell *optionCell = NULL;
+
+	/* Extract options from the statement node tree */
+	foreach(optionCell, copyStatement->options)
+	{
+		DefElem *option = (DefElem *) lfirst(optionCell);
+
+		/* skip master related options */
+		if ((strncmp(option->defname, "master_host", NAMEDATALEN) == 0) ||
+			(strncmp(option->defname, "master_port", NAMEDATALEN) == 0))
+		{
+			continue;
+		}
+
+		newOptionList = lappend(newOptionList, option);
+	}
+
+	copyStatement->options = newOptionList;
+}
+
+
+static char
+RemotePartitionMethod(char *relationName, PGconn *connection)
+{
+	char partitionMethod = '\0';
+	PGresult *queryResult = NULL;
+
+	StringInfo partitionMethodCommand = makeStringInfo();
+	appendStringInfo(partitionMethodCommand, PARTITION_METHOD_QUERY, relationName);
+
+	queryResult = PQexec(connection, partitionMethodCommand->data);
+	if (PQresultStatus(queryResult) == PGRES_TUPLES_OK)
+	{
+		char *partitionMethodString = PQgetvalue((PGresult *) queryResult, 0, 0);
+		partitionMethod = partitionMethodString[0];
+	}
+	else
+	{
+		ReportRemoteError(connection, queryResult);
+		ereport(ERROR, (errmsg("could not get the partition method of the "
+							   "distributed table")));
+	}
+
+	PQclear(queryResult);
+
+	return partitionMethod;
+}
+
+
+static void
+UpdateRemoteShardStatistics(Oid relationId, uint64 shardId, PGconn *connection)
+{
+	PGresult *queryResult = NULL;
+	StringInfo updateShardStatisticsCommand = makeStringInfo();
+	char *relationName = NULL;
+
+	/* if there is not a remote connection, fall back to local logic */
+	if (connection == NULL)
+	{
+		UpdateShardStatistics(relationId, shardId);
+		return;
+	}
+
+	relationName = get_rel_name(relationId);
+	appendStringInfo(updateShardStatisticsCommand, UPDATE_SHARD_STATISTICS_QUERY,
+					 relationName, shardId);
+
+	queryResult = PQexec(connection, updateShardStatisticsCommand->data);
+	if (PQresultStatus(queryResult) != PGRES_TUPLES_OK)
+	{
+		ereport(WARNING, (errmsg("couldn't update shard statistics"),
+						  errhint("You can update shard statistics by manually "
+								  "running master_update_shard_statistics() udf.")));
+	}
+
+	PQclear(queryResult);
+}
+
+
+static List *
+RemoteFinalizedShardPlacementList(uint64 shardId, PGconn *connection)
+{
+	PGresult *queryResult = NULL;
+	StringInfo shardPlacementsCommand = makeStringInfo();
+	List *finalizedPlacementList = NIL;
+
+	/* if there is not a remote connection, fall back to local logic */
+	if (connection == NULL)
```

Comment:
> I think it would be better if there was an intermediate function that switches to either local or remote logic, also because this part doesn't seem to match the function name.

Naming is bit tricky here, currently, I named the intermediate function as
CallFinalizedShardPlacementList

other options I can think of

DispatchFinalizedShardPlacementList
LocalOrRemoteFinalizedShardPlacementList
GetFinalizedShardPlacementList

Do you have any suggestions?